pkg/scheduler/queue/query_component_load.go

```diff
@@ -0,0 +1,187 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package queue
+
+import (
+	"math"
+	"sync"
+
+	"github.com/pkg/errors"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+)
+
+type QueryComponent string
+
+const (
+	StoreGateway QueryComponent = "store-gateway"
+	Ingester     QueryComponent = "ingester"
+)
+
+var QueryComponents = []QueryComponent{
+	StoreGateway,
+	Ingester,
+}
+
+// cannot import constants from frontend/v2 due to import cycle
+// these are attached to the request's AdditionalQueueDimensions by the frontend.
+const ingesterQueueDimension = "ingester"
+const storeGatewayQueueDimension = "store-gateway"
+const ingesterAndStoreGatewayQueueDimension = "ingester-and-store-gateway"
+
+// QueryComponentNameForRequest parses the expected query component from annotations by the frontend.
+func QueryComponentNameForRequest(req *SchedulerRequest) string {
+	var expectedQueryComponent string
+	if len(req.AdditionalQueueDimensions) > 0 {
+		expectedQueryComponent = req.AdditionalQueueDimensions[0]
+	}
+
+	return expectedQueryComponent
+}
+
+// queryComponentFlags interprets annotations by the frontend for the expected query component,
+// and flags whether a query request is expected to be served by the ingesters, store-gateways, or both.
+func queryComponentFlags(queryComponentName string) (isIngester, isStoreGateway bool) {
```

Comment:
If I understand this correctly, this takes in a queue dimension (`ingester`, `store-gateway` or `ingester-and-store-gateway`), not a single component name (`ingester` or `store-gateway`). Perhaps `queueDimension` would be a clearer name rather than `queryComponentName`?

Reply:
same idea as above - I want this `queueDimension` concept to go away because it's not accurate or really useful.